### PR TITLE
feat(memory): expose agent semantic file memory as a search_knowledge_base source (ISSUE-256)

### PR DESCRIPTION
**Pull request by basnijholt**
## Summary

Agent file memory was already searchable through `search_memories`, but the agent-scoped semantic file-memory index never appeared as a source in `search_knowledge_base`. An agent reasoning about "what can I search?" from its own tool descriptions concluded that its memory was unsearchable.

This exposes a ready agent-scoped semantic file-memory index as a **read-only** `search_knowledge_base` source, while keeping `search_memories` as the CRUD-capable surface.

## Root cause

File-memory knowledge bases were created as a runtime overlay (`_semantic_file_search.py`) and never entered the agent's configured knowledge-base id list. `_semantic_agent_knowledge_base_ids()` in `knowledge/utils.py` built the searchable set only from `config.resolve_entity(agent_name).knowledge_base_ids`, so the overlay id never reached `resolve_agent_knowledge_access()`, `KnowledgeWithSourceDescriptions`, or the generated tool description.

## Changes

**Stage 0 — description and docs accuracy**

- `search_memories` description now distinguishes configured semantic include patterns from the fixed `MEMORY.md` + `memory/**/*.md` keyword/fallback corpus, and distinguishes CRUD-capable persisted / `file:*` IDs from read-only `semantic:*` locators.
- The generated knowledge-tool description was made non-closed-world so it no longer implies its source list covers every searchable corpus.
- Deleted the nonexistent `resolve_knowledge_owner_bindings` Tach interface exposure. No shim.

**Stage 1 — expose the index**

- New cycle-free leaf `src/mindroom/file_memory_knowledge.py` owns the file-memory base ID, include patterns, and runtime overlay construction. These builders were **moved** out of `_semantic_file_search.py`, not copied, so memory and knowledge callers share one base and one indexing signature.
- `resolve_agent_knowledge_access()` appends the runtime memory base after authored semantic bases and sends ready handles through the existing merge path.
- `create_agent()` enables knowledge capabilities whenever runtime resolution produced a non-`None` knowledge handle.
- Canonical `agent_<name>` scope derivation moved to a cycle-free `memory_scope_ids.py` leaf shared by both memory and knowledge paths.

## Isolation and safety

- Per-agent isolation is enforced: two genuinely distinct, non-symlinked agent workspaces publish different base IDs and reject cross-agent retrieval **in both directions**.
- Private-agent resolution stays fail-closed without identity; distinct requesters receive distinct roots and base IDs.
- Router resolution returns before runtime materialization.
- Cold runtime memory indexes cannot publish while their managed corpus is empty; scheduler ownership is claimed atomically across scheduler instances.
- The content-publication gate is persisted in index metadata and participates in corpus compatibility, preventing an older ungated empty index from being reused.

## Preserved behavior (explicitly tested)

- Semantic degradation to keyword results, including the classified `degraded_reason`.
- System-prompt memory auto-injection.
- `search_memories` CRUD contract: `semantic:*` IDs are read-only locators, proven by a search-to-CRUD test where `get` returns no record and `update`/`delete` reject those locators.
- Docker projection: runtime-only overlays do not enter authored worker projection.

## Review

Seven independent reviews of `cdecdf75b` converged on seven defects or cleanup items, all addressed at their owning boundaries rather than patched at call sites.

## Validation

- Full repository suite passed on the final tree, excluding only the pre-existing `tests/test_ci_release_dispatch.py` NixOS failure (that test replaces `PATH` with `/usr/bin:/bin`; NixOS provides no `/bin/bash`).
- `uv run tach check --dependencies --interfaces` passed.
- All pre-commit hooks passed except the platform-specific `ty` hook, which cannot resolve macOS-only `AppKit` / `ApplicationServices` / `Quartz` in untouched desktop modules on NixOS. Direct `ty` checks over every changed Python file passed.

### Live test

Ran against a disposable lab instance via `nix-shell`, all six assertions passing:

```
PASS distinct-workspaces: alpha and beta use separate non-symlink roots
PASS cold-notice: file_memory_agent_alpha_... stayed unpublished while its managed corpus was empty
PASS notice-cleared: no file_memory_* availability notice remained after publication
PASS search_knowledge_base: alpha marker returned from memory/notes.md
PASS search_memories: alpha marker returned with a read-only semantic locator
PASS cross-agent-negative: beta exposed no alpha marker through either search surface
RESULT ISSUE-256 LIVE TEST: PASS
```

An earlier live run surfaced a real bug: the new publication-gate identity field was omitted from persisted metadata, so a successful refresh reloaded as `CONFIG_MISMATCH`. Fixed at the metadata serialization / corpus-compatibility boundary and covered by a red/green unit regression.

## Blast radius

The production config was loaded read-only and evaluated through `config.resolve_entity()`. All 16 configured agents resolve to `memory_backend=file` and `memory_search.mode=semantic`, so all of them gain the new read-only source. No service was restarted or modified as part of this work.

29 files changed, 982 insertions, 130 deletions.

## Summary by Sourcery

Expose per-agent semantic file-memory indexes as read-only knowledge sources and align memory/knowledge tooling, isolation, and indexing behavior around a shared overlay.

New Features:
- Add a shared file-memory knowledge overlay module that derives stable base IDs, roots, and configs for agent-scoped semantic file memory.
- Expose ready agent file-memory indexes through resolve_agent_knowledge_access so search_knowledge_base can query configured file memory as a semantic-only source.
- Enable knowledge search for agents with only file-backed semantic memory by wiring knowledge capabilities when any knowledge handle is available.

Bug Fixes:
- Prevent duplicate knowledge refreshes across multiple scheduler instances by atomically claiming refresh targets.
- Gate publication of cold runtime file-memory indexes until managed content exists, avoiding empty indexes and reusing incompatible ungated metadata.
- Ensure semantic file-memory search IDs are treated as read-only locators and rejected by CRUD operations.
- Fix knowledge-source merging to allocate enough result budget to include appended memory sources.
- Preserve router behavior by avoiding runtime resolution for non-agent names when resolving file-memory knowledge.
- Align indexing metadata so the new content publication gate participates in corpus compatibility and does not misclassify refreshed indexes.

Enhancements:
- Refactor semantic file-memory indexing to use a shared, cycle-free file_memory_knowledge module and memory_scope_ids helpers for scope IDs.
- Update knowledge search tool descriptions to clearly distinguish search_knowledge_base sources from memory search capabilities and conditionally recommend search_memories.
- Improve search_memories tool documentation to clarify corpus behavior, semantic fallback, and CRUD semantics for different ID types.
- Clarify keyword vs semantic file-memory corpus rules and semantic fallback behavior in user-facing memory and tools documentation, including OpenClaw guidance.

Build:
- Register the new file_memory_knowledge and memory_scope_ids modules and their dependencies in tach.toml, exposing them to the knowledge and memory subsystems.

Documentation:
- Update memory and tools documentation to describe semantic file-memory search as a read-only knowledge source, clarify corpus and fallback semantics, and distinguish CRUD-capable IDs from semantic locators.
- Extend OpenClaw docs to explain how file memory participates in knowledge search and when to prefer search_memories over search_knowledge_base.

Tests:
- Add focused tests for the shared file-memory knowledge overlay, covering shared identity, agent resolution constraints, and private-agent requester isolation.
- Add integration tests that validate file-memory overlays are queryable via knowledge search, remain unpublished when empty, and are isolated across agents.
- Add tests confirming semantic memory search IDs are read-only for CRUD, keyword search uses a fixed corpus, and degraded semantic searches avoid semantic-only paths.
- Add tests ensuring knowledge search tool descriptions reflect memory tool availability and that memory-only agents receive appropriate semantic knowledge source descriptions.
- Add tests for the refresh scheduler’s exclusivity across instances and for the content publication gate’s impact on indexing settings corpus keys.

## Summary by CodeRabbit

* **New Features**
  * File-backed semantic memory can now be searched through the knowledge base for eligible agents.
  * Added an option to delay publishing empty semantic indexes until content is available.
  * Search results now clearly distinguish editable memory IDs from read-only semantic locators.

* **Bug Fixes**
  * Prevented duplicate knowledge refreshes from running concurrently.
  * Improved memory search fallback and scope isolation.

* **Documentation**
  * Clarified keyword versus semantic search, supported identifiers, CRUD behavior, fallback rules, and when to use each memory search tool.

### CODE REVIEW

File: src/mindroom/config/knowledge.py [hunk 1]
```diff
@@ -79,6 +79,10 @@ class KnowledgeBaseConfig(BaseModel):
         default=True,
         description="When true, shared local folders watch filesystem changes and schedule background published-index refresh without blocking reads; when false, direct external file edits require explicit reindex or dashboard/API mutations",
     )
+    require_content_before_publish: bool = Field(
+        default=False,
+        description="Keep a cold semantic index initializing until at least one managed source file exists",
+    )
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/config/knowledge.py:85]: Fixed in a9efa719f: documented require_content_before_publish in docs/knowledge.md and docs/configuration/index.md, with generated skill references refreshed.

File: src/mindroom/file_memory_knowledge.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    from mindroom.constants import RuntimePaths
+    from mindroom.tool_system.worker_routing import ToolExecutionIdentity
+
+_CHUNK_SIZE = 5000
+_CHUNK_OVERLAP = 0
+_MEMORY_KNOWLEDGE_PREFIX = "file_memory"
+_FILE_MEMORY_DESCRIPTION = (
+    "Configured file memory for this agent. "
+    "Read-only semantic search over configured Markdown paths (default: memory/**/*.md)."
+)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/file_memory_knowledge.py:27]: Fixed in a9efa719f: added file_memory_knowledge.py and memory_scope_ids.py to CLAUDE.md and docs/architecture/index.md, with generated references refreshed.

File: src/mindroom/knowledge/utils.py [hunk 3]
```diff
@@ -714,7 +724,7 @@ def _merge_knowledge(agent_name: str, knowledges: list[Knowledge]) -> Knowledge
     return KnowledgeWithSourceDescriptions(
         name=f"{agent_name}_multi_knowledge",
         vector_db=_MultiKnowledgeVectorDb(vector_dbs=vector_db_sources),
-        max_results=max(knowledge.max_results for knowledge in queryable_knowledges),
+        max_results=max(len(queryable_knowledges), *(knowledge.max_results for knowledge in queryable_knowledges)),
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/knowledge/utils.py]: Fixed in a9efa719f: bounded the source-count-derived merged default at 20 while preserving explicitly configured per-source defaults. Regression covers 10 sources and a 25-source cap.

File: src/mindroom/knowledge/refresh_scheduler.py [hunk 4]
```diff
@@ -194,7 +194,9 @@ def _start_task(self, key: KnowledgeRefreshTarget, request: _ScheduledRefresh) -
         loop = _running_loop_for_schedule(key.base_id)
         if loop is None:
             return
-        mark_refresh_active(key)
+        if not try_mark_refresh_active(key):
+            logger.debug("Skipping duplicate knowledge refresh owned by another scheduler", base_id=key.base_id)
+            return
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/knowledge/refresh_scheduler.py]: Fixed in a9efa719f: scheduler claims now distinguish scheduler duplicates from direct refresh owners. Direct-owner-blocked requests retain the latest request and retry after release; cross-scheduler duplicates remain deduplicated. Both paths have regressions.
